Update metadata

AbeJellinek opened the pull request:
Supersedes #1582 (because I can't force-push to mrtcode/zotero). This is basically @mrtcode's work plus my tweaks from #1582, squashed to make rebasing less terrible, and updated to work on Z7. Ready for design review and bug testing.

AbeJellinek committed:
Rate-limit URL translation by host

Including resolved DOI URLs.

And move metadata updating utilities out of utilities_internal.
They deserve their own file, especially since only method one is
actually called from outside.

AbeJellinek committed:
Update metadata

Co-authored-by: Martynas Bagdonas <martbgd@gmail.com>

AbeJellinek committed:
Rough Z7 update

AbeJellinek committed:
Move DTD strings to Fluent

AbeJellinek committed:
Move window title to Fluent

And remove seemingly unused general.updated and general.result.

dstillman commented:
We have an older version of Diff Match and Patch from https://github.com/zotero/zotero/commit/a3654652b21071955ae01ba8fcdc38765e1740a4. Let's just update that one. (We can check whether it still works when we get back to https://github.com/zotero/zotero/pull/3468.)

AbeJellinek committed:
Consolidate diff.js and diff-match-patch.js

Use the newer version of the library from diff-match-patch.js, but
remove match/patch functions since we don't need them either.

AbeJellinek committed:
Use accurate naming in diff_wordsToChars_()

AbeJellinek committed:
diff_wordsToChars_(): Treat each whitespace as its own word

This gives us nicer diffs without trailing whitespace.

Also rewrite the iterative splitting method using split(). split()
doesn't actually double the memory footprint on any modern JS engine,
since it creates dependent strings, not copies:
    https://searchfox.org/mozilla-central/rev/5c2888b35/js/src/builtin/String.cpp#3704-3705
And the iterative method was hard to read and maintain.

dstillman commented:
Just starting to look at this, but just from the first 30 seconds:

![Screenshot 2025-03-18 at 2 47 33 PM](https://github.com/user-attachments/assets/eff9cf8c-31e4-4caa-977a-34afa4304627)

1. This Abstract diff is…not usable. Do we have any control over this?
2. Done vs. Apply All is definitely not going to work. I'll let @yexingsha make some suggestions here, but we need to structure the expected interaction flow better, and I think we also want to optimize for the almost certainly default case of wanting to select all items in your library, selecting Update (or its better future name), and having it update all items to whatever you would get with the magic wand, completely overwriting anything you have, without thinking about it. We want to give the people who want more fine-grained control more control and try to avoid data catastrophes (wiping out custom plugin data in Extra? deleting abstracts that were added manually?), but most people are just going to want an automatic process.

dstillman commented:
((2) probably also has implications for rate limiting, progress display, and maybe even caching. I would expect people to repeatedly select all items and re-run this.)

AbeJellinek committed:
Move styles to SCSS

And really prevent resizing (maybe temporarily, until we figure out the
layout).

AbeJellinek committed:
Ignore meaningless spacing changes in abstract

And use correct field name for abstract checks.

yexingsha commented:
> "Update" feels too short in the context menu, but "Update Metadata" is too technical and "Update Item" is weird (and goes against our resolution to stop putting "Item" in every menu item). We could do something a little longer but more descriptive, like "Search for Updates"? Or "Find Updates" to match "Find Full Text"?

I think we can call it "Update Info", since "Info" is the name for the metadata section in ItemPane.

I agree that we should optimize for the use case of "updating multiple items without thinking about it". When there are multiple items being updated, all details should be collapsed by default; when there is only one item being updated, then we can expand it by default. The heuristics we use to deselect certain updates does complicate things a bit, though I can see the utility in e.g. not applying destructive changes. I think we should expose these heuristics as options, such as "update everything", "do not delete existing fields", etc.

I made some designs in [figma](https://www.figma.com/design/y8aZzEDwsxsgVh7Zui63to/Zotero?node-id=12828-138820&t=sK8sHnrBU6xLYekG-1) exploring how best to represent the updates when all details are collapsed. This is the approach I'm most happy with:

<img width="900" alt="in-progress" src="https://github.com/user-attachments/assets/b04494b4-4938-4304-a63f-094913982022" />

<img width="900" alt="finished" src="https://github.com/user-attachments/assets/d2aef992-fd48-46bb-9c19-37b9d5373436" />

<img width="900" alt="expanded" src="https://github.com/user-attachments/assets/60f20e03-c264-47c2-84db-1af3698e56bf" />

- The main action buttons "Cancel" and "Apply Updates" are moved to the bottom right. I removed the buttons for each individual item since they seem redundant. I added "Expand All" and "Collapse All" to the bottom left for easier individual selections.
- In the default collapsed state, you can choose to read the list of updates and expand items to modify them, or simply ignore them and apply whatever updates your chose option provides. Some further considerations:
    - How to display fields where updates are available but not selected. Here I chose to display them with a strikethrough.
    - Whether to include the second line of details when there is no update. I chose to put "No Update" there for now.
    - When there are a lot of updated fields, whether to wrap the list into multiple lines. I guess cutting it off would be too arbitrary?
- The expanded state is similar to the current implementation, with a few changes:
    - I think right now it's not clear enough that clicking on a diff will select/unselect it, so I added checkmarks before each field.
    - Let's restrict all diffs to one line, since you can always expand them.
    - Changed background color to `tag-red` and `accent-green` at 50% opacity.

yexingsha commented:
> "Update" setting could be hard to follow sometimes. For instance, if the item type changes, we might *need* to clear some fields.

I think we need to describe the update heuristic as much as we can. Something like "Default" would be too vague and confusing. What about "Avoid Clearing Existing Fields"?

> In this design, do you still expand the diff by clicking the ellipsis? Now that we have a checkbox for each field, maybe we can remove the click-to-toggle functionality and make a click anywhere on the diff expand it instead?

Yes, that's a good idea.

> Should items have their own checkboxes so you can quickly choose not to apply any of the updates it's suggested? Overriding and disabling the field-specific checkboxes?

That is a good point. I tried it and found that checkboxes don't go well with the item type icons, so I just got rid of the icons:

<img width="900" alt="in-progress" src="https://github.com/user-attachments/assets/1098d9f2-1d6a-40aa-9df6-61138eea1eaa" />

<img width="900" alt="finished" src="https://github.com/user-attachments/assets/5eaf0f6d-1c08-4813-8045-54983643f730" />

<img width="900" alt="expanded" src="https://github.com/user-attachments/assets/116a7346-1783-49dc-8283-20e5dab76f5f" />

This also means items without update will stand out more. What do you think?

dstillman commented:
For controlling the clearing of existing fields, a checkbox might be more straightforward than radio buttons (and we might have other options at some point)?

yexingsha commented:
> For controlling the clearing of existing fields, a checkbox might be more straightforward than radio buttons (and we might have other options at some point)?

Then I guess it would be a checkbox labeled "Avoid Clearing Existing Fields" that's checked by default. Though maybe it gives too little warning about how dangerous unchecking it would be?

AbeJellinek committed:
WIP: Implement new design

AbeJellinek committed:
Add border

AbeJellinek committed:
Improve toggle keyboard access

AbeJellinek committed:
fixup

dstillman commented:
The URL value is appearing a line down from the label:

![Screenshot 2025-04-17 at 11 45 35 PM](https://github.com/user-attachments/assets/ea6c4aff-85a4-437c-a800-862dc929a1c4)

And, actually, I'm seeing it for some other fields too:

![Screenshot 2025-04-17 at 11 52 16 PM](https://github.com/user-attachments/assets/01b139be-114a-42f3-be8c-f22294cf9870)

I'm also not totally clear what's going on with the truncation. This is what I get if I click on the above URL — it definitely had space:

![Screenshot 2025-04-17 at 11 45 43 PM](https://github.com/user-attachments/assets/95a530c5-23b1-4c3d-9bc2-861c4cf80d60)

dstillman commented:
Closing the window with the red stoplight button on macOS doesn't cancel the process. The icon in the top right keeps spinning, and if I then try to update info for another item, it reopens the pane with all the previous items that I cancelled.

dstillman commented:
> Maybe we should keep Cancel around, so closing with the red stoplight won't clear the queue, but reopening it and then explicitly clicking Cancel will.

OK, yes, that would work. I think we definitely want a get-me-out-of-here button for an operation like this.

AbeJellinek committed:
Don't hide Cancel

AbeJellinek committed:
Restore cleanup step

AbeJellinek committed:
Increase truncation threshold
